clippy_lints/src/unnecessary_map_on_constructor.rs

```diff
@@ -0,0 +1,96 @@
+}
+declare_lint_pass!(UnnecessaryMapOnConstructor => [UNNECESSARY_MAP_ON_CONSTRUCTOR]);
+
+impl<'tcx> LateLintPass<'tcx> for UnnecessaryMapOnConstructor {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx rustc_hir::Expr<'tcx>) {
+        if expr.span.from_expansion() {
+            return;
+        }
+        if let hir::ExprKind::MethodCall(path, recv, args, ..) = expr.kind {
+            let is_option = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Option);
+            let is_result = is_type_diagnostic_item(cx, cx.typeck_results().expr_ty(recv), sym::Result);
+
+            if let (false, false) = (is_option, is_result) {
+                return;
+            };
+
+            let (constructor_path, constructor_item) = if_chain! {
+                if let hir::ExprKind::Call(constructor, constructor_args) = recv.kind;
+                if let hir::ExprKind::Path(constructor_path) = constructor.kind;
+                if let Some(arg) = constructor_args.get(0);
+                then {
+                    (constructor_path, arg)
+                } else {
+                    return;
+                }
+            };
+            let constructor_symbol = match constructor_path {
+                hir::QPath::Resolved(_, path) => {
+                    if let Some(path_segment) = path.segments.last() {
+                        path_segment.ident.name
+                    } else {
+                        return;
+                    }
+                },
+                hir::QPath::TypeRelative(_, path) => path.ident.name,
+                hir::QPath::LangItem(_, _, _) => return,
+            };
+            match constructor_symbol {
+                sym::Some | sym::Ok if path.ident.name == rustc_span::sym::map => (),
+                sym::Err if path.ident.name == sym!(map_err) => (),
+                _ => return,
+            }
+
+            if_chain! {
+                if let Some(arg) = args.get(0);
+                if let hir::ExprKind::Path(fun) = arg.kind;
+                then {
+                    let mut applicability = Applicability::MachineApplicable;
+                    let fun_snippet = snippet_with_applicability(cx, fun.span(), "_", &mut applicability);
+                    let constructor_snippet =
+                        snippet_with_applicability(cx, constructor_path.span(), "_", &mut applicability);
+                    let constructor_arg_snippet =
+                        snippet_with_applicability(cx, constructor_item.span, "_", &mut applicability);
+                    let suggestion = format!("{constructor_snippet}({fun_snippet}({constructor_arg_snippet}))");
+                    span_lint_and_sugg(
+                        cx,
+                        UNNECESSARY_MAP_ON_CONSTRUCTOR,
+                        expr.span,
+                        &format!("unnecessary {fun_snippet} on contstuctor {constructor_snippet}(_)"),
+                        "try using",
```

Comment:
```suggestion
                        &format!("unnecessary `{}` on constructor `{constructor_snippet}(_)`", path.ident.name),
                        "try",
```

Reply:
I've removed the "using", but path.ident.name does not point to the `map` argument but to `map` itself, I've kept the `fun_snippet` for that reason

Reply:
Yeah that's intended, the `map`/`map_err` is the unnecessary part, `fun_snippet` is necessary it's just moved around a bit

Reply:
ok, I didn't realize that was for the warning message. After d501c3b
they and the solution proposed should be correct:

> warning: unnecessary map on constructor Ok(_)

> help: try: `Ok(fun(x))`